File: problems-101-150/p145.c
#include<stdio.h>

long reverse(long n) {
	long x = n;
	long y = 0;

	while (x) {
		y *= 10;
		y += (x % 10);
		x /= 10;
	}

	return y;
}

int reversible(long n) {
	long m = reverse(n);
	long sum = m + n;
	int rev = 1;

	while (sum) {
		if ((sum % 10) % 2 == 0) {
			rev = 0;
			break;
		}
		sum /= 10;
	}

	return rev;
}

int main() {
	long i, total = 0;

	for (i = 0; i < 1000000000; i++) {
		if ((i % 10) && reversible(i)) total++;
	}
	
	printf("Done! \n%ld\n\n", total);
	return 0;
}

File: problems-101-150/p145.hs
-- This approach is too slow. Switching to Python...
import Data.Char
import Data.List

digits :: Integer -> [Int]
digits x = map digitToInt (show x)

-- "reverse" is a built-in name
rev :: Integer -> Integer
rev x = read (reverse $ show x) :: Integer

reversible :: Integer -> Bool
reversible x = all odd (digits (x + (rev x)))

main = do
    let candidates = filter (\x -> x `mod` 10 /= 0) [1,2..1000]
    let revs = filter reversible candidates 
    putStrLn $ show revs

